kobo/client/__init__.py

```diff
@@ -216,7 +244,12 @@ def _login(self, force=False, verbose=False):
 
         # create new self._hub instance (only once, when calling constructor)
         if self._hub is None:
-            self._hub = xmlrpclib.ServerProxy("%s/%s/" % (self._hub_url, self._client_type), allow_none=True, transport=self._transport, verbose=verbose)
+            success, message = self.check_hub_connectivity(self._hub_url)
```

Comment:
Given that this will completely block initialization when it returns False, I'm a little nervous about a couple of the details of how it works.

Firstly, it doesn't use the same `transport` that the `self._hub` itself uses.  This means there's some behavior differences.  I don't necessarily know all the differences, but at least the SSL verification behavior will be different, I believe.  Because you just use `requests.get` without any particular SSL-related options, this change *probably* has the side-effect of suddenly requiring the hub's cert to verify against whatever CA bundle the requests library is using, while it didn't previously have this requirement.  (If you check the initialization of `self._transport`, it seems like it doesn't actually verify certs by default - unless the caller has set `CA_CERT` config option - but if they have set that, the new code added here doesn't respect it, so at the very least that's a bug.)

Secondly, there's the fact that at least some login methods are allowed and expected to have some side-effect on `self._transport`, e.g. adding a session cookie into the cookie jar.  Doing this check *before* that happens means not seeing that side-effect yet.  If the server is configured to, say, always require auth for that endpoint or else give a 401 or 403 response, then the check prior to login isn't going to work.  I'm not sure it's safe to assume that it won't be configured that way, seeing as kobo doesn't own the web server config.

Is there a way to write this so that it's not adding a new HTTP-level check which works differently from most requests?  Can we just use the XML-RPC client "normally" and somehow make the resulting errors more meaningful to end-users?

Reply:
I tried various approaches earlier, including passing in the `_transport`. I failed to find a way to verify the accessibility/validaty of the endpoint without initializing a `HubProxy` instance along with transport/auth.

The approach attempts to fix one issue but indeed poses various other problems. I'll see if there's a better way, and I'll close this PR, thanks